feat: add holding and swipe player gestures

United600 opened the pull request:
Refactors the gesture handling system for the media player, making it more flexible and configurable. Instead of using simple boolean flags for gestures like volume and seek, it now allows mapping various gestures (tap and swipe directions) to specific media commands via a new `PlaybackActionKind` enum. This improves extensibility and allows for more customized user interactions.

Adds press and hold gesture support (boosts playback speed)

It also refactors the `VideoView.PointerWheelChanged` and `VlcVideoView.SizeChanged` methods to the view, and adds support for horizontal touchpad/mouse wheel input.

<img width="703" height="496" alt="Gesture settings" src="https://github.com/user-attachments/assets/b284ca16-6db2-4d0d-8df2-308875ab9bf5" />

<img width="703" height="496" alt="Gesture settings with the options expanded" src="https://github.com/user-attachments/assets/5b64a7a4-f338-4b85-9fa8-aefcd5d9c18c" />

United600 committed:
refactor gesture system and UI

United600 committed:
add tap and hold gesture and refactor geture handling to use recognizer

United600 committed:
rename gesture methods and simplify the playback speed login in player element viewmodel

United600 committed:
refactor player element mouse wheel handling and add support for seeking

United600 committed:
refactor gesture options to use array instead of list

United600 committed:
simplify vertical and horizontal gesture logic

United600 committed:
improve holding gesture to avoid redundant playback speed changes

United600 committed:
refactor player view size handling to code-behind

United600 committed:
use delta for volume changes on pointer wheel changes

United600 committed:
update gesture settings headers resource string names

United600 committed:
improve enum resource converter fallback behavior

United600 committed:
refactor play pause gesture handling

United600 committed:
Revert "refactor play pause gesture handling"

This reverts commit 87209090fd7965d5ede398206cbd315f40db279c.

United600 committed:
fix tap gesture conflict with double click

United600 committed:
refactor method order in player element view model

United600 committed:
Merge branch 'main' into United600/gestures

United600 committed:
Merge branch 'main' into United600/gestures

United600 committed:
add combobox behavior to gesture related comboboxes

United600 committed:
update pointer event handlers process element

United600 committed:
refactor swipe gesture handling and add settings for two-finger gestures

United600 committed:
Merge branch 'main' into United600/gestures

United600 committed:
rename player gesture enum

United600 committed:
improve enum to resource string conversion logic

United600 committed:
remove support for two-finger slide gestures

United600 committed:
refactor and simplify gesture handling

United600 committed:
add missing behavior to the swipe right combobox setting

United600 committed:
remove manipulation lock enum and add return early for none gesture option

United600 committed:
Merge branch 'main' into United600/gestures

United600 committed:
Merge branch 'main' into United600/gestures

United600 committed:
refactor player gesture logic

rename player gesture enum

add new action for playback speed control

United600 committed:
refactor pointer wheel input to use fixed step values

improve playback rate formatting and gesture cleanup

minor converter xml doc update

United600 committed:
rename gesture properties for consistency

update playback rate handling and enum converter

United600 committed:
Merge branch 'main' into United600/gestures

United600 committed:
improve holding status message to persist

fix ci building error

United600 committed:
refactor status message to use duration instead of boolean

United600 committed:
add hold space key to speed up

update keyboard accelerator helper

United600 committed:
refactor enum to resource string converter

United600 committed:
refactor playback rate formatting to humanizer class

United600 committed:
respect press and hold gesture setting for space key holding

United600 committed:
minor xml docs improvements

United600 committed:
Merge branch 'main' into United600/gestures

United600 committed:
restore slide gestures

United600 committed:
lock slide axis direction

United600 committed:
minor tweak to gesture settings resources

huynhsontung commented on Screenbox.Core/ViewModels/SettingsPageViewModel.cs:
```diff
@@ -58,6 +63,8 @@ public sealed partial class SettingsPageViewModel : ObservableRecipient
 
     public int[] PlayerControlsHideDelayOptions { get; } = { 1, 2, 3, 4, 5 };
 
+    public Array GestureOptions { get; }
```
This is a weird type. You meant `object[]`?

United600 replied:
No, I just copied the return type from the Community Toolkit [EnumValuesExtensions](https://github.com/CommunityToolkit/Windows/blob/main/components/Extensions/src/Markup/EnumValuesExtension.cs). If you prefer we could've use the extension to set the `Enum` directly as the `ItemSource`.

huynhsontung committed:
concrete type for GestureOptions array

huynhsontung committed:
adjust seeking amount for slide gesture and horizonal mouse wheel

huynhsontung committed:
reduce wait time for press and hold

huynhsontung committed:
handle some edge cases with touch and some minor refactor

United600 commented on Screenbox/Controls/PlayerElement.xaml.cs:
```diff
@@ -53,4 +85,69 @@ private void VideoViewButton_OnDoubleTapped(object sender, DoubleTappedRoutedEve
+
+    private void VideoViewButton_OnPointerMoved(object sender, PointerRoutedEventArgs e)
+    {
+        if (!IsEnabled) return;
```
```suggestion
        if (!IsEnabled || ViewModel.IsHolding) return;
```

United600 commented on Screenbox.Core/ViewModels/PlayerPageViewModel.cs:
```diff
@@ -290,9 +293,66 @@ public void OnPointerMoved()
+            }
+
+            _isSpaceKeyHolding = true;
+        }, TimeSpan.FromMilliseconds(300));
```
I couldn't find official documentation specifying the default keyboard repeat delay, but 500 ms seemed consistent with the behavior of the other inputs (touch and mouse).

huynhsontung committed:
restore holding ignore pointer moved events

This is intentional to allow the user to move their pointer while holding and still keeping the hold gesture active.

Co-authored-by: United600 <United600@users.noreply.github.com>

huynhsontung committed:
restore the hold space key delay to 500ms

huynhsontung committed:
tune horizontal scrolling

huynhsontung committed:
adjust how _timeBeforeManipulation should be handled
to report accurate delta time after gestures

huynhsontung commented on Screenbox.Core/Services/SettingsService.cs:
```diff
@@ -38,6 +35,13 @@ public sealed class SettingsService : ISettingsService
     private const string PlayerShowChaptersKey = "Player/ShowChapters";
     private const string PrivacyPersistPlaybackPosition = "Privacy/PersistPlaybackPosition";
 
+    private const string PlayerGestureTapKey = "Player/Gesture/Tap";
```
Casting `bool` to `int` is planned. True value will be cast to 1, which matches `PlaybackActionKind.PlayPause`. Maintain backward compatibility.

United600 commented on Screenbox.Core/Enums/PlaybackActionKind.cs:
```diff
@@ -0,0 +1,22 @@
+/// <summary>
+/// Represents the action that was taken on the media playback.
+/// </summary>
+public enum PlaybackActionKind
```
Should we consider converting this enum to a flag enum (and renaming it to `PlaybackActionKinds`) now? This would allow us to add pointer wheel (and slide gesture) customization, though we'd need to update the `ResourceNameToResourceStringConverter` but that shouldn't be too difficult.

Then we could do something like this:

```csharp
// SettingsPageViewModel.cs
public PlaybackActionKind[] PointerWheelOptions { get; }

...

var pointerWheelSeekAction = PlaybackActionKind.Rewind | PlaybackActionKind.FastForward;
var pointerWheelVolumeAction = PlaybackActionKind.DecreaseVolume | PlaybackActionKind.IncreaseVolume;

PointerWheelOptions = new[]
{
    PlaybackActionKind.None,
    pointerWheelSeekAction,
    pointerWheelVolumeAction,
};

// PlayerElementViewModel.cs
private void ProcessPlayerGesture(PlaybackActionKind gestureOption, double change)
{
    ...

    switch (gestureOption)
    {
        ...

        case PlaybackActionKind.Rewind | PlaybackActionKind.FastForward:
            ...
            break;
        case PlaybackActionKind.DecreaseVolume | PlaybackActionKind.IncreaseVolume:
            ...
            break;
    }
}
```

huynhsontung replied:
I'm not sure I see the benefits of making this enum a flag. Especially when it's user facing since it involves resource string. Should the slide gesture use the same enum? Maybe it warants its own enum since the behavior of slide gesture is fundamentally different from other gestures. Even if we were to use the same enum then can't we add dedicated entries to the enum instead of using calculated ones?

United600 replied:
> Should the slide gesture use the same enum? Maybe it warants its own enum since the behavior of slide gesture is fundamentally different from other gestures. Even if we were to use the same enum then can't we add dedicated entries to the enum instead of using calculated ones?

Are they really that different? My thinking is along the lines of keeping all the gesture-related logic within the `ProcessPlayerGesture` method, but once we ship this there's no turning back with that enum.

> I'm not sure I see the benefits of making this enum a flag. Especially when it's user facing since it involves resource string.

I don't think this will be a major problem. We can check for combined values, allowing us to use a resource string like this:

```xml
<data name="PlaybackActionKindDecreaseVolumeIncreaseVolume" xml:space="preserve">
  <value>Change volume</value>
</data>
```

United600 replied:
> Even if we were to use the same enum then can't we add dedicated entries to the enum instead of using calculated ones?

Sure, though it wouldn't be as "elegant" and would require updating the `GestureOptions` in the `SettingsViewModel` (which isn't a major issue).

huynhsontung replied:
I still don't think we should make it a flag. I still hold to my recommendation that we either create a new enum or add dedicated values to the existing `PlaybackActionKind`.

United600 replied:
I guess it's fine, and we avoid some complexity when dealing with the converter. Just wanted to make sure we don't regret it later.

huynhsontung committed:
restore enum to track slide gesture direction

huynhsontung commented on Screenbox/Strings/en-US/Resources.resw:
```diff
@@ -664,13 +664,10 @@
-  </data>
-  <data name="SettingsGestureTap" xml:space="preserve">
-    <value>Tap anywhere to play or pause</value>
+    <value>Enable slide vertically to adjust volume</value>
```
These changes our language a little bit. Adding "Enable" at the start sounds too programmatic. Maybe because we have two verbs next to each other? "Enable slide...". I prefer not having "Enable..."

United600 replied:
I took inspiration from this system setting, but it's okay to remove it. Do you think we also should change the press and hold string to something like "Press and hold temporarily increases the playback speed", or "Press and hold to temporarily increase the playback speed"? The second option seems off without "Enable...".

<img width="1544" height="920" alt="Windows 11 additional touch settings" src="https://github.com/user-attachments/assets/6826b329-0cfb-4d34-8c7c-79c52f098ac9" />

huynhsontung replied:
"Press and hold to temporarily increase the playback speed" sounds off to me, too. "Press and hold temporarily increases the playback speed" is also not technically correct English, either. Maybe for press and hold, we follow the system settings and add "Enable..." at the front. Keep the slide strings without.

United600 replied:
So, "**Enable press and hold to increase playback speed**", or "**Enable press and hold to temporarily increase playback speed**"? Does "press and hold" clearly indicate that the action is temporary, or should we include "temporarily"? I think we can omit it.

Other alternative, "Enable press and hold to speed up playback".

huynhsontung replied:
I think we can skip "temporarily". So "Enable press and hold to increase playback speed" is good to me 👍

United600 commented on Screenbox.Core/ViewModels/PlayerElementViewModel.cs:
```diff
@@ -30,6 +29,17 @@ public sealed partial class PlayerElementViewModel : ObservableRecipient,
 {
+    private const double GestureStepAmount = 5.0;
+
+    private enum SlideDirection
```
```suggestion
    private enum SlideAxis
```

To me, direction means left, right, up, down. `SlideRail`? `SlideOrientation`? I think "Axis" and "Rail" lean more toward "X/Y/Z", so "Orientation" might be the better choice.

huynhsontung replied:
I agree. `SlideOrientation` is good for me. Good catch.

United600 replied:
🤔 We could make `_slideOrientation` nullable and use the platform [Orientation](https://learn.microsoft.com/uwp/api/windows.ui.xaml.controls.orientation) enum. Maybe not, but I should've used that instead of a nullable bool.

huynhsontung replied:
No, that's fine. We are not saving much by reusing an enum :)

United600 committed:
rename playback action kind template for clarity

United600 committed:
rename slide enum and simplify settings strings
